### PR TITLE
feat: Optimize slot finding answer processing

### LINKED ISSUE #307: Entities in response are not selected by accuracy
**Describe the bug**
If several occurrences of the same entity are found in an utterance, the answer seems to use the one with the lowest position.

**To Reproduce**
```js
const { NlpManager, ConversationContext } = require('node-nlp');

async function main() {
  const manager = new NlpManager({ languages: ['en'] });
  const ctx = new ConversationContext();

  manager.addNamedEntityText('style', 'rock', ['en'], ['rock', 'Rock']);

  manager.slotManager.addSlot('music', 'style', true, {
    en: 'What style of music do you want?',
  });
  manager.addDocument('en', 'Play me some music', 'music');
  // This next line does not change anything but it's something I tried
  manager.addDocument('en', 'I want some %style%', 'music');
  manager.addAnswer('en', 'music', "Ok let's play some {{style}}");

  await manager.train();
  await manager.process('en', 'Play me some music', ctx);
  const result = await manager.process('en', 'I want some rock', ctx);
  console.log(JSON.stringify(result, null, 2));
}

main();
```

Here the log of the last process response. You can see one entity with a perfect accuracy smashed by the next one:
```json
{
  "utterance": "I want some rock",
  "locale": "en",
  "languageGuessed": false,
  "localeIso2": "en",
  "language": "English",
  "domain": "default",
  "classifications": [
    {
      "label": "music",
      "value": 1
    }
  ],
  "intent": "music",
  "score": 1,
  "entities": [
    {
      "start": 12,
      "end": 15,
      "len": 4,
      "levenshtein": 0,
      "accuracy": 1,
      "option": "rock",
      "sourceText": "rock",
      "entity": "style",
      "utteranceText": "rock"
    },
    {
      "entity": "style",
      "utteranceText": "I want some rock",
      "sourceText": "I want some rock",
      "accuracy": 0.95,
      "start": 0,
      "end": 15,
      "len": 16
    }
  ],
  "sourceEntities": [],
  "sentiment": {
    "score": 0.475,
    "comparative": 0.11875,
    "vote": "positive",
    "numWords": 4,
    "numHits": 4,
    "type": "senticon",
    "language": "en"
  },
  "actions": [],
  "srcAnswer": "Ok let's play some {{style}}",
  "answer": "Ok let's play some I want some rock"
}
```

**Expected behavior**
The entity used in answers should be the one with the higher accuracy value.

**Desktop (please complete the following information):**
 - OS: macOS 10.14.3
 - node: 10.15.1
 - axa-nlp: 3.10.1

**Additional context**
The problem seems to be here https://github.com/axa-group/nlp.js/blob/master/lib/nlp/nlp-manager.js#L451 and probably here too https://github.com/axa-group/nlp.js/blob/master/lib/nlp/nlp-manager.js#L433

### LINKED ISSUE #582: Utterances with duplicated Entities in v4
**Describe the bug**
We're using NlpManager to match intents and extract named and builtin entities from utterances. It works as expected when we don't have duplicated entities. The problem arrises when we have more than 1 entity of the same type in one sentence. To identify entities, we use the following format {entity}_{integer} (it was explained [here](https://github.com/axa-group/nlp.js/blob/master/docs/v3/ner-manager.md#utterances-with-duplicated-entities)). However, after processing the message, we only retrieve the plain entities (e.g., `hero`) without the suffix number (e.g., expected `here_1`). Maybe I'm missing something. Could you help us solve this issue please?

**To Reproduce**

```
const { NlpManager } = require('node-nlp');

async function main() {
    const manager = new NlpManager({ languages: ['en'], nlu: { useNoneFeature: false }});
     manager.addDocument(
        'en',
        'I saw %hero_1% together with %hero_2%, they were eating %food%',
        'saw_heroes_eating'
    );

    manager.addNamedEntityText('hero', 'spiderman', ['en'], ['Spider-man']);
    manager.addNamedEntityText('hero', 'iron man', ['en'], ['iron man']);
    manager.addNamedEntityText('hero', 'thor', ['en'], ['Thor']);
    manager.addNamedEntityText('food', 'burger', ['en'], ['Burger']);
    manager.addNamedEntityText('food', 'pizza', ['en'], ['pizza']);
    manager.addNamedEntityText('food', 'pasta', ['en'], ['Pasta', 'spaghetti']);

    await manager.train();

    const result = await manager.process('en','I saw spiderman together with ironman, they were eating spaghetti');
    console.log(result);
}

main();
```
The result is:
```
{ locale: 'en',
  utterance:
   'I saw spiderman together with ironman, they were eating spaghetti',
  languageGuessed: false,
  localeIso2: 'en',
  language: 'English',
  nluAnswer:
   { classifications: [ [Object], [Object] ],
     entities: undefined,
     explanation: undefined },
  classifications:
   [ { intent: 'saw_heroes_eating', score: 1 },
     { intent: 'sum_numbers', score: 0 } ],
  intent: 'saw_heroes_eating',
  score: 1,
  domain: 'default',
  sourceEntities: [],
  entities:
   [ { start: 6,
       end: 14,
       len: 9,
       levenshtein: 1,
       accuracy: 0.9,
       entity: 'hero',
       type: 'enum',
       option: 'spiderman',
       sourceText: 'Spider-man',
       utteranceText: 'spiderman' },
     { start: 30,
       end: 36,
       len: 7,
       levenshtein: 1,
       accuracy: 0.875,
       entity: 'hero',
       type: 'enum',
       option: 'iron man',
       sourceText: 'iron man',
       utteranceText: 'ironman' },
     { start: 56,
       end: 64,
       len: 9,
       levenshtein: 0,
       accuracy: 1,
       entity: 'food',
       type: 'enum',
       option: 'pasta',
       sourceText: 'spaghetti',
       utteranceText: 'spaghetti' } ],
  answers: [],
  answer: undefined,
  actions: [],
  sentiment:
   { score: 1.625,
     numWords: 10,
     numHits: 3,
     average: 0.1625,
     type: 'senticon',
     locale: 'en',
     vote: 'positive' } }
```

**Expected behavior**
`entity` value should be `hero_1` or `hero_2`
```
{ locale: 'en',
  utterance:
   'I saw spiderman together with ironman, they were eating spaghetti',
  languageGuessed: false,
  localeIso2: 'en',
  language: 'English',
  nluAnswer:
   { classifications: [ [Object], [Object] ],
     entities: undefined,
     explanation: undefined },
  classifications:
   [ { intent: 'saw_heroes_eating', score: 1 },
     { intent: 'sum_numbers', score: 0 } ],
  intent: 'saw_heroes_eating',
  score: 1,
  domain: 'default',
  sourceEntities: [],
  entities:
   [ { start: 6,
       end: 14,
       len: 9,
       levenshtein: 1,
       accuracy: 0.9,
       entity: 'hero_1',
       type: 'enum',
       option: 'spiderman',
       sourceText: 'Spider-man',
       utteranceText: 'spiderman' },
     { start: 30,
       end: 36,
       len: 7,
       levenshtein: 1,
       accuracy: 0.875,
       entity: 'hero_2',
       type: 'enum',
       option: 'iron man',
       sourceText: 'iron man',
       utteranceText: 'ironman' },
     { start: 56,
       end: 64,
       len: 9,
       levenshtein: 0,
       accuracy: 1,
       entity: 'food',
       type: 'enum',
       option: 'pasta',
       sourceText: 'spaghetti',
       utteranceText: 'spaghetti' } ],
  answers: [],
  answer: undefined,
  actions: [],
  sentiment:
   { score: 1.625,
     numWords: 10,
     numHits: 3,
     average: 0.1625,
     type: 'senticon',
     locale: 'en',
     vote: 'positive' } }
```
**Screenshots**

**Desktop (please complete the following information):**
 - version 4.10.7
 - Node version 10.16.3

### PR DESCRIPTION
**Pull request by Apollon77**
# Pull Request Template

## PR Checklist

- [x] I have run `npm test` locally and all tests are passing.
- [x] I have added/updated tests for any new behavior.
- [x] If this is a significant change, an issue has already been created where the problem / solution was discussed: [N/A, or add link to issue here]

I currently can not intall the project because of issues reported with some used tarballs. WHen I execute the tests eslint reports 16 errors and 47 warnings all over the place. I can try to fix them, but So I decided to not add new tests. Please tell me if needed

## PR Description
While checking out slot Filling with v4 I encountered several unexpected topics I solve in this PR:
* When using slot filling there was a "srcAnswer" property added to the result, but this was not processed like the other answers by nlg. The PR adds that
* the nlg text rendering had an issue that when a contact was set that then "optional answer text" was not processed because done but ignored and original text was used. PR also addresses that
* The answers were rendered before the slot filling filled relevant entities. This PR adjust the order of actions to allow entities to be rendered correctly
* When a slot was requested the full uterance was always filled as entity - also if the entity was already parsed by buoltins or such. This lead to duplicated entries. This PR changes it so that the fallback "full utterance" is only added as entity if not filled already
* If a context contains a earlier srcAnswer property then this was repeated also if all mandatory entities were filled in the meantime. This PR removes the srcAnswer if all mandatory entities are filled.
* If slot filling is used the nlp code adds a special "entity utterance", but uses the initial locale parameter (whcih could be empty) to create this. This pot. leads to an empty locale used and then ner only uses the default builtins and not the configured builtins (e.g, Duckling). This PR uses a potentially already determinded locale.
* When slot filling generated an answer this was provided in srcAnwer additionally to pot. a normal "answer". DirectLineConnector as example is not sending back srcAnswer, so the question is not sent out. This PR changes the behaviour so that when the srcAnswer property is set this is also used as "answer", because it is "the better answer" :-)
* Actions are executed after the answers are rendered. With this actions can not modify the context to e.g. inkject new entities to be used in answer rendering. This PR adds a new nlp setting "executeActionsBeforeAnswers" to allow to execute the actions before answer rendering.
* Fix handling of numbered entities in slot filling and answer generation (added it here because else would have conflicted with PR #1164)

In fact it is a feature addition for handing slot filling answers by processing them like the other answers.

addresses #410, addresses #454, addresses #934, fixes #582, fixes #307

Additonally the commits are not "convenionally named". Sorry for that. Saw that too late and will change in future commits

### CODE REVIEW

**Comment by Apollon77** [comment]
PS: The npm ci issues can be seens here too https://github.com/Apollon77/nlp.js/runs/7564017521?check_suite_focus=true#step:4:50

**Comment by Apollon77** [comment]
PPS: The issue is that the Artifactory Server that is used as packagelocatiopn for some packages is not activated licensing wise anymore as it seems ... It redirects to https://my.jfrog.com/reactivate-server/axags ... it seems that some local packages were used instead official ones.

I will prepare a second PR with updated package-locks and fixed tests

**Comment by Apollon77** [comment]
see https://github.com/axa-group/nlp.js/pull/1168

Once the abobe is merged I could add also tests here if required

File: packages/nlp/test/nlp.test.js [hunk 1]
```diff
@@ -788,6 +788,97 @@ describe('NLP', () => {
...
+      const input = {
+        locale: 'en',
+        utterance: 'Who am i?',
+      };
+      const output = await nlp.process(input);
+      expect(output.utterance).toEqual(input.utterance);
+      expect(output.intent).toEqual('who_am_i');
+      expect(output.answer).toEqual('answer');
+    });
+    test('The action is executed before answer generation (when configured that way) and answer is not overridden', async () => {
```
Comment (eric-lara) [inline packages/nlp/test/nlp.test.js:857]: This test is almost equal than the 2nd you added. If you want to test executeActionsBeforeAnswers add cases who clearly explain the difference between setting it to true, false &  validate that using default value is being understood as false value.
Basically, with value true allow answers to be rendered with entities filled from slot filling. With false, show how this fails.

Comment (Apollon77) [reply packages/nlp/test/nlp.test.js:857]: Yes you are right ... I will rework that

File: packages/slot/src/slot-manager.js [hunk 2]
```diff
@@ -210,12 +239,12 @@ class SlotManager {
...
       });
-    }
-    for (let i = 0, l = result.entities.length; i < l; i += 1) {
-      delete mandatorySlots[result.entities[i].entity];
+      delete mandatorySlots[context.slotFill.currentSlot];
     }
     keys = Object.keys(mandatorySlots);
     if (!keys || keys.length === 0) {
+      // All mandatory slots are filled, so we are done. No further questions needed
+      delete result.srcAnswer;
```
Comment (eric-lara) [inline packages/slot/src/slot-manager.js:247]: As far as I know, the purpose of srcAnswer was to keep track of the raw answer without being processed at all.

Example:
answer:  "Ok let's play some rock"
srcAnswer:  "Ok let's play some {{style}}"

Comment (Apollon77) [reply packages/slot/src/slot-manager.js:247]: In this case we only talk about slot filling quertions and not "In general".

result.srcAnswer is initialized with the last slotfill answer aka "question" from the last run (line 205 this file). And was in the past the also preserved when the latest user input resolved all open manatory slots. This then had effects and the same question as last time was potentially asked again by slot filling algorithm.

This then had the effect that because srcAnswer was still existing, but not relevant anymore.

When digging through code I can not find other places where srcAnswer is used than for slotfilling.

In fact your idea makes a lot of sense and could be added. The past code have set srcAnswer in slot filling but Never replaced contained entities at all ... I added that :) So I change that noew here to have srcAnswer as the raw text. Good idea

File: packages/nlp/test/nlp.test.js (hunk 1 above)
Comment (Apollon77) [reply packages/nlp/test/nlp.test.js:857]: Ahh I see it now ... tests did not followed all feature additions :-) I added the option that a answer set in an action executed before determine answers is used and the answer determination is skipped. I fix the tests

Comment (Apollon77) [reply packages/nlp/test/nlp.test.js]: tests adjusted and enhanced to show the power of the new feature :-)

**Comment by Apollon77** [comment]
Ready for re-review

File: packages/slot/src/slot-manager.js (hunk 2 above)
Comment (Apollon77) [reply packages/slot/src/slot-manager.js:247]: PS: if you would like to introduce srcAnswer as field in general also for the normal answers, this can be done with changes in nlg and nlp. Should be straight forward.

**Comment by ericzon** [comment]
ieps! it's complaining due to [max-classes-per-file rule](https://github.com/axa-group/nlp.js/actions/runs/3569213924/jobs/5998939614#step:6:78). 

TemplateMock appear 2 times in [nlp.test.js](https://github.com/axa-group/nlp.js/blob/d579236c67b4a978ecaaaa48451bfc7181aeff63/packages/nlp/test/nlp.test.js). As it's exactly the same class, can be fixed quickly, leaving just a single implementation.

File: packages/slot/src/slot-manager.js (hunk 2 above)
Comment (ericzon) [reply packages/slot/src/slot-manager.js:247]: we can leave it like this by now but have it on mind for next updates, thanks!

**Comment by Apollon77** [comment]
Sw that too but was in bed already ... I move to an extra file now ... stay tuned

**Comment by Apollon77** [comment]
> TemplateMock appear 2 times in [nlp.test.js](https://github.com/axa-group/nlp.js/blob/d579236c67b4a978ecaaaa48451bfc7181aeff63/packages/nlp/test/nlp.test.js). As it's exactly the same class, can be fixed quickly, leaving just a single implementation.

BTW: Ws not because of that ... it was the general "more then one class per file" error. Solved now

**Comment by Apollon77** [comment]
Re Sonarcloud Bug ... this is a false positive now ...

Bildschirm­foto 2022-11-29 um 07 46 11

Or what I oversee? (yes ok I could check for null explicitely, but falsy does the same in the end

**Comment by ericzon** [comment]
The problem is because SonarCloud is having on mind the jsDoc of the method:
image

**Comment by Apollon77** [comment]
hehe .. ok ... will adjust it

**Comment by Apollon77** [comment]
Heeyyy he is fine with it

PS: If you like you can "Squash&Merge" this PR because I think this commit history migth not be that helpful at the end ;-)

**Comment by ericzon** [comment]
node v16 ci got stucked... I'll rerun later

**Comment by ericzon** [comment]
It gets stuck in tests of v16 per 2nd time. I'll trigger them again keeping debug logs. If this fails, we'll have to run it on local to see what's going on.

**Comment by Apollon77** [comment]
Strange ... I have nodejs 16.17.0 locally ... running it now completely

**Comment by Apollon77** [comment]
hhmmm ... I get a oom issue ... let me try again

**Comment by ericzon** [comment]
Finally! running only failed jobs I was able to pass v16 job 👍

**Comment by Apollon77** [comment]
Cool ... I lso did not got oom again on second run but two other checks failed that are fixed in master already because of the other PR ... so that also might be a topic because I did not rebased on master

**Comment by ericzon** [comment]
well, as all the tests passed here and we're going to merge more PRs soon, let's proceed
